### PR TITLE
Last line of defense, best effort serialization

### PR DESCRIPTION
**Pull request by BadSingleton**
### What does this implement/fix? Explain your changes.

This commit adds a "last line of defense, best effort serialization"
to serialize types not marked as Serializable. Such objects are
deserialized as derived classes with all methods and properties
overriden to throw a "Not Serialized" Exception.
Fields are not initialized and may be null.

Private classes and other non-inheritable instances are set to null on deserialization

### Does this close any currently open issues?

...

### Any other comments?

...

### Checklist

Check all those that are applicable and complete.

-   [ ] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [ ] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [ ] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

File: src/runtime/StateSerialization/RuntimeData.cs [hunk 1]
```diff
@@ -3,18 +3,303 @@
...
 using Python.Runtime.StateSerialization;
 
 using static Python.Runtime.Runtime;
 
 namespace Python.Runtime
 {
+    [System.Serializable]
+    public sealed class NotSerializedException: SerializationException
+    {
+        static string _message = "The underlying C# object has been deleted.";
```
Comment (benoithudson) [inline src/runtime/StateSerialization/RuntimeData.cs:20]: should this be const?

File: src/runtime/StateSerialization/RuntimeData.cs [hunk 2]
```diff
@@ -3,18 +3,303 @@
...
+                                    new[] { propertyType });
+
+                GenerateExceptionILCode(setPropMthdBldr);
+                propertyBuilder.SetSetMethod(setPropMthdBldr);
+            }
+        }
+
+        private static void CreateMethod(TypeBuilder tb, MethodInfo minfo)
+        {
+            Console.WriteLine($"overimplementing method for: {minfo}  {minfo.IsVirtual}  {minfo.IsFinal}  ");
```
Comment (benoithudson) [inline src/runtime/StateSerialization/RuntimeData.cs]: probably meant to remove this

File: src/runtime/StateSerialization/RuntimeData.cs [hunk 3]
```diff
@@ -3,18 +3,303 @@
...
+            // object.
+            if (obj.GetType().Name == "TypeLoadExceptionHolder")
+            {
+                obj = null!;
+                return;
+            }
+
+            MaybeType type = obj.GetType();
+            
+            var hasAttr = (from attr in obj.GetType().CustomAttributes select attr.AttributeType == typeof(PyNet_NotSerializedAttribute)).Count() != 0;
```
Comment (benoithudson) [inline src/runtime/StateSerialization/RuntimeData.cs]: Maybe `type.CustomAttributes.Any(attr => attr.AttributeType == typeof(...))` ?

I don't know the fancy way to write it, but you don't need to count it, you can stop at the first.

(And you already got obj.GetType())

File: src/runtime/StateSerialization/RuntimeData.cs [hunk 4]
```diff
@@ -3,18 +3,303 @@
...
+            if (hasAttr)
+            {
+                // Don't serialize a _NotSerialized. Serialize the base type, and deserialize as a _NotSerialized
+                type = type.Value.BaseType;
```
Comment (benoithudson) [inline src/runtime/StateSerialization/RuntimeData.cs:336]: I take it we know that a NotSerialized always derives from a serializable object?

File: src/runtime/StateSerialization/RuntimeData.cs [hunk 5]
```diff
@@ -3,18 +3,303 @@
...
+                // test: the object successfully serialized, but it cannot be
+                // deserialized.
+                // just return null.
+                return null!;
+            }
+
+            object nameObj = null!;
+            try
+            {
+                nameObj = info.GetValue($"notSerialized_tp", typeof(object));
```
Comment (benoithudson) [inline src/runtime/StateSerialization/RuntimeData.cs]: you don't need to interpolate this string, there's no fields inside

File: src/runtime/StateSerialization/RuntimeData.cs [hunk 6]
```diff
@@ -3,18 +3,303 @@
...
+                throw new ArgumentNullException();
+            }
+            if (type.IsSerializable)
+            {
+                selector = this;
+                return null; // use whichever default
+            }
+            else
+            {
+                selector = this;
```
Comment (benoithudson) [inline src/runtime/StateSerialization/RuntimeData.cs]: both branches do it so you could do it before the if/else

**Review by benoithudson — APPROVED** [review]
Awesome.

Should we backport to the 3.0 alpha tag as well?

File: src/runtime/StateSerialization/RuntimeData.cs (hunk 4 above)
Comment (BadSingleton) [reply src/runtime/StateSerialization/RuntimeData.cs:336]: No, it always derives from a non-serializable object. That's why we end up here. We don't serialize the object; only it's type so we can re-create it in the next domain.

File: src/runtime/StateSerialization/RuntimeData.cs (hunk 2 above)
Comment (BadSingleton) [reply src/runtime/StateSerialization/RuntimeData.cs]: yes

File: src/runtime/StateSerialization/RuntimeData.cs (hunk 1 above)
Comment (BadSingleton) [reply src/runtime/StateSerialization/RuntimeData.cs:20]: Is this even the right message? This is a twisted joke I made to trigger PTSD of all the poor souls who had to work with PyQT/PySide.

**Comment by BadSingleton** [comment]
> Should we backport to the 3.0 alpha tag as well?

Considering that RC.1 has been declared the other day, I'll have to take a look to see if we can use it instead of the alpha

File: src/runtime/StateSerialization/RuntimeData.cs (hunk 3 above)
Comment (BadSingleton) [reply src/runtime/StateSerialization/RuntimeData.cs]: Oh, that's better

File: src/runtime/StateSerialization/RuntimeData.cs (hunk 1 above)
Comment (benoithudson) [reply src/runtime/StateSerialization/RuntimeData.cs:20]: It's a perfect message.

**Pull request merged**